New module: Add module for Bugzilla (web_infrastructure/bugzilla_bug)

bhavikbhavsar opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
Ansible Module to fetch details of particular bugzilla from bugzilla server
<!---
If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - New Module Pull Request


##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
Bugzilla
##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
```
ansible 2.4.0.0
```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```

```

bhavikbhavsar committed:
Ansible Module for Bugzilla

bhavikbhavsar committed:
Ansible Module for Bugzilla added ANSIBLE_METADATA and RETURN

bhavikbhavsar committed:
fixed errors caught by ansibot #22514

bhavikbhavsar committed:
fixed support_by to community #22514

bhavikbhavsar committed:
removing as keyword to make it compatible with python 2.4

bhavikbhavsar commented:
@mattclay Could you please suggest what changes I should make so that it's compatible with both PY2 and PY3? I have converted my code to be compatible with python 2.4 .

mattclay commented:
@bhavikbhavsar See the documentation on module [exceptions](https://docs.ansible.com/ansible/dev_guide/developing_modules_python3.html#exceptions).

bhavikbhavsar committed:
adding sys.exec_info to fix compatibility issues

bhavikbhavsar committed:
remove import xmlrpclib since it is not required in exception

bhavikbhavsar commented:
@Slezhuk @inetfuture @jhoekx @jtyr @lekum @mattupstate @mgruener @n0trax @oboukili @ramondelafuente @robinro @scottanderson42 @sermilrod @tarka @tastychutney

Could you please review and submit your vote ?

jtyr commented on lib/ansible/modules/web_infrastructure/bugzilla.py:
```diff
@@ -0,0 +1,145 @@
+# Created on Mar 10, 2017
+# @author: Bhavik Bhavsar (9.bhavik@gmail.com)
+#
+#
```
Here should be only one blank line.

jtyr commented on lib/ansible/modules/web_infrastructure/bugzilla.py:
```diff
@@ -0,0 +1,145 @@
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+#
+
+ANSIBLE_METADATA = {'status': ['preview'],
```
I would format it like this:

```
ANSIBLE_METADATA = {
    'status': ['preview'],
    'supported_by': 'community',
    'version': '1.0'
}
```

jtyr commented on lib/ansible/modules/web_infrastructure/bugzilla.py:
```diff
@@ -0,0 +1,145 @@
+module: bugzilla
+short_description: Perform various actions with bugzilla
+description:
+    - Fetch status of particular bugzilla id from Bugzilla url provided.
```
Standard indentation is 2 spaces. Please correct it in all `DOCUMENTATION` block.

jtyr commented on lib/ansible/modules/web_infrastructure/bugzilla.py:
```diff
@@ -0,0 +1,145 @@
+
+EXAMPLES = '''
+
+#for private bug
```
Please format it like this:

```
# For private bug
```

No empty line after that comment.

jtyr commented on lib/ansible/modules/web_infrastructure/bugzilla.py:
```diff
@@ -0,0 +1,145 @@
+    get_status: True
+  register: status
+
+#for non-private bug
```
The same as above.

jtyr commented on lib/ansible/modules/web_infrastructure/bugzilla.py:
```diff
@@ -0,0 +1,145 @@
+
+- name: Get status for BZ 1429535
+  bugzilla:
+    url: "bugzilla.redhat.com"
```
You don't need to quote the string.

jtyr commented on lib/ansible/modules/web_infrastructure/bugzilla.py:
```diff
@@ -0,0 +1,145 @@
+  bugzilla:
+    url: "bugzilla.redhat.com"
+    bug_id: 1429535
+    get_status: True
```
This should be:

```
get_status: yes
```

jtyr commented on lib/ansible/modules/web_infrastructure/bugzilla.py:
```diff
@@ -0,0 +1,145 @@
+    bug_id: 14295
+    user: <username>
+    password: <password>
+    get_status: True
```
This should be:

```
 get_status: yes
```

jtyr commented on lib/ansible/modules/web_infrastructure/bugzilla.py:
```diff
@@ -0,0 +1,145 @@
+  bugzilla:
+    url: "bugzilla.redhat.com"
+    bug_id: 14295
+    user: <username>
```
Please give an example of the `username` value.

jtyr commented on lib/ansible/modules/web_infrastructure/bugzilla.py:
```diff
@@ -0,0 +1,145 @@
+    url: "bugzilla.redhat.com"
+    bug_id: 14295
+    user: <username>
+    password: <password>
```
Please give an example of the `password` value.

jtyr commented on lib/ansible/modules/web_infrastructure/bugzilla.py:
```diff
@@ -0,0 +1,145 @@
+
+- name: Get status for BZ 14295
+  bugzilla:
+    url: "bugzilla.redhat.com"
```
This string should not be quoted.

jtyr commented on lib/ansible/modules/web_infrastructure/bugzilla.py:
```diff
@@ -0,0 +1,145 @@
+  description: Fetches status of bugzilla
+  returned: Only when commands get_status is set to True and Bug_id is provided
+  type: string
+  sample: "NEW"
```
The string value should not be quoted.

jtyr commented on lib/ansible/modules/web_infrastructure/bugzilla.py:
```diff
@@ -0,0 +1,145 @@
+RETURN = '''
+status:
+  description: Fetches status of bugzilla
+  returned: Only when commands get_status is set to True and Bug_id is provided
```
Please change `True` to `C(yes)` and `Bug_id` to `I(bug_id)`.

jtyr commented on lib/ansible/modules/web_infrastructure/bugzilla.py:
```diff
@@ -0,0 +1,145 @@
+        required: true
+    user:
+        description:
+            - username to login into Bugzilla
```
Change `username` to `I(username)`.

jtyr commented on lib/ansible/modules/web_infrastructure/bugzilla.py:
```diff
@@ -0,0 +1,145 @@
+options:
+    url:
+        description:
+            - url of bugzilla server
```
Please change `url` to `URL`.

jtyr commented on lib/ansible/modules/web_infrastructure/bugzilla.py:
```diff
@@ -0,0 +1,145 @@
+version_added: "2.3"
+author: Bhavik Bhavsar
+requirements:
+    - "python >= 2.6"
```
The string doesn't have to be quoted. The same bellow.

jtyr commented on lib/ansible/modules/web_infrastructure/bugzilla.py:
```diff
@@ -0,0 +1,145 @@
+    get_status = module.params['get_status']
+
+    if not bug_id:
+        module.fail_json(msg="bug_id missing")
```
Not necessary as the `bug_id` is marked as `required` so the module framework will check for it itself.

bhavikbhavsar replied:
@jtyr: But when the parameter is defined and value is not defined it would give error as below
`int() argument must be a string or a number, not 'NoneType'` 
is there someway where module framework checks for value is present ?

jtyr replied:
This is exactly the kind of check the module framework is doing. It's telling you that the parameter is required and it must be a `string` (which gets translated to `int`, I guess) or an `int` and that the value it sees is of type `None`. The point here is that the custom error message you defined is never triggered because the module checks the value before it reaches your condition.

jtyr commented on lib/ansible/modules/web_infrastructure/bugzilla.py:
```diff
@@ -0,0 +1,145 @@
+        module.fail_json(msg="bug_id missing")
+
+    if not get_status:
+        module.fail_json(msg="get_status param not set")
```
This should be not required if the `get_status` has some default value.

jtyr commented on lib/ansible/modules/web_infrastructure/bugzilla.py:
```diff
@@ -0,0 +1,145 @@
+            bug_id = dict(required=True, type='int'),
+            user = dict(required=False, type='str'),
+            password = dict(required=False, type='str', no_log=True),
+            get_status = dict(required=False, type='bool')
```
This should have default value of `true`.

jtyr commented on lib/ansible/modules/web_infrastructure/bugzilla.py:
```diff
@@ -0,0 +1,145 @@
+
+    except:
+        t, e = sys.exc_info()[:2]
+        response = {e.faultCode:e.faultString}
```
Please format this like this:

```
response = {
    e.faultCode: e.faultString
}
```

jtyr commented on lib/ansible/modules/web_infrastructure/bugzilla.py:
```diff
@@ -0,0 +1,145 @@
+author: Bhavik Bhavsar
+requirements:
+    - "python >= 2.6"
+    - "python-bugzilla >= 2.0.0"
```
Wouldn't be better to do the API query instead of adding additional dependency?

jtyr commented on lib/ansible/modules/web_infrastructure/bugzilla.py:
```diff
@@ -0,0 +1,145 @@
+#!/usr/bin/python
+#
+# Created on Mar 10, 2017
+# @author: Bhavik Bhavsar (9.bhavik@gmail.com)
```
The above two lines should be formated like this:

```
# (c) 2017, Bhavik Bhavsar <9.bhavik@gmail.com>
```

jtyr commented on lib/ansible/modules/web_infrastructure/bugzilla.py:
```diff
@@ -0,0 +1,145 @@
+    try:
+        bzapi = bugzilla.Bugzilla(url, user, password)
+        bug = bzapi.getbug(bug_id)
+
```
There should be no blank line here.

jtyr commented on lib/ansible/modules/web_infrastructure/bugzilla.py:
```diff
@@ -0,0 +1,145 @@
+        t, e = sys.exc_info()[:2]
+        response = {e.faultCode:e.faultString}
+        module.fail_json(msg="FaultError", FaultError=response)
+
```
There should be no blank line here.

jtyr commented on lib/ansible/modules/web_infrastructure/bugzilla.py:
```diff
@@ -0,0 +1,145 @@
+    else:
+        status = bugzilla_status(bug)
+        module.exit_json(changed=False, status=status)
+
```
There should be two blank lines here.

jtyr reviewed (changes requested):
Please use `flake8` tool to review your code before you submit it.

bhavikbhavsar committed:
fixed changes suggested in code review

Signed-off-by: bbhavsar <bbhavsar@redhat.com>

bhavikbhavsar committed:
minor fix missed W292 no newline at end of file

bhavikbhavsar committed:
moved import below RETURN and changed version to 2.4

jtyr commented:
The module name should be something less generic (e.g. bugzilla_bug) as there might be other modules modifying other things in Bugzilla (e.g adding/updating users, comments, ...). The module interface is weird. The module should allow to create and update the bug. The `get_status` doesn't feel right. It should use classical `state: present` instead and the module should return details of the bug that can be then accessed via a registered variable.

You should also consider to introduce the `params` option as used in the `yum_repository` and the `jenkins_plugin` modules.

gsr-shanks commented:
> The module name should be something less generic (e.g. bugzilla_bug) as there might be other modules modifying other things in Bugzilla (e.g adding/updating users, comments, ...).

IMO this module has the capability to support other things in Bugzilla as well. I don't think we should be creating multiple modules for the same web infrastructure.

jtyr commented:
@gsr-shanks It's always better to have smaller modules than one big.

gsr-shanks commented:
@jtyr I looked at other Jenkins modules in web_infrastructure namespace which are smaller ones. That said, we have a generic Jira module as well. But I agree that smaller modules are better.

bhavikbhavsar commented:
Hi @jtyr 

I am trying to 1st convert my code-logic to not have dependencies from python-bugzilla. 
I have tried this logic with basic requests library.

For using it with ansible's library I am using below import
`from ansible.module_utils.urls import *`

and logic as 
```
    querystring = {"method":"Bug.get","params":"[{\"ids\": [%s]} ]"%(bug_id)}
    full_uri = url + "/jsonrps.cgi"
    resp = open_url(full_uri, data=querystring, headers={'Content-Type':'application/json'}, method="GET")
    result = resp.read()
    jsonresult = json.loads(result)
    req.close()
    return jsonresult
```
For some reason when I try to execute it fails with

```
The full traceback is:
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/ansible/executor/task_executor.py", line 125, in run
    res = self._execute()
  File "/usr/lib/python2.7/site-packages/ansible/executor/task_executor.py", line 525, in _execute
    result = self._handler.run(task_vars=variables)
  File "/usr/lib/python2.7/site-packages/ansible/plugins/action/normal.py", line 45, in run
    results = merge_hash(results, self._execute_module(tmp=tmp, task_vars=task_vars, wrap_async=wrap_async))
  File "/usr/lib/python2.7/site-packages/ansible/plugins/action/__init__.py", line 636, in _execute_module
    (module_style, shebang, module_data, module_path) = self._configure_module(module_name=module_name, module_args=module_args, task_vars=task_vars)
  File "/usr/lib/python2.7/site-packages/ansible/plugins/action/__init__.py", line 160, in _configure_module
    task_vars=task_vars, module_compression=self._play_context.module_compression)
  File "/usr/lib/python2.7/site-packages/ansible/executor/module_common.py", line 806, in modify_module
    (b_module_data, module_style, shebang) = _find_module_utils(module_name, b_module_data, module_path, module_args, task_vars, module_compression)
  File "/usr/lib/python2.7/site-packages/ansible/executor/module_common.py", line 688, in _find_module_utils
    recursive_finder(module_name, b_module_data, py_module_names, py_module_cache, zf)
  File "/usr/lib/python2.7/site-packages/ansible/executor/module_common.py", line 514, in recursive_finder
    msg.append('either %s.py or %s.py' % (py_module_name[-1], py_module_name[-2]))
IndexError: tuple index out of range

fatal: [localhost]: FAILED! => {
    "failed": true, 
    "msg": "Unexpected failure during module execution.", 
    "stdout": ""
}
```

I believe this is known issue [#16470](https://github.com/ansible/ansible/issues/16470)
Hence I tried below upgrades in the virtualenv
```
pip install --upgrade pip 
pip install --upgrade setuptools
pip install 'git+https://github.com/ansible/ansible.git@devel'
```

Could you please point me in correct direction ?

jtyr commented:
@bhavikbhavsar You should use the `fetch_url` method which wraps the `open_url` for you. It will also allow you to use all common module options like `url_username`, `url_password`. Please have a look at the `jenkins_plugin` module how it should be used.

bhavikbhavsar committed:
1. Removed dependency for python-bugzilla library
2. Added params style similar to jenkins plugin
3. used fetch_url method as per review comment

Signed-off-by: bbhavsar <bbhavsar@redhat.com>

bhavikbhavsar committed:
fixed pep8 rules and changed version to 2.4

bhavikbhavsar commented:
Hi @jtyr  could you please review the changes ?
I have removed the dependency for python-bugzilla library and also made changes as per the jenkins plugin module.

jtyr commented on lib/ansible/modules/web_infrastructure/bugzilla_bug.py:
```diff
@@ -0,0 +1,160 @@
+version_added: "2.4"
+author: Bhavik Bhavsar
+requirements:
+    - "python >= 2.6"
```
Isn't this the default requirement for Ansible now?

bhavikbhavsar replied:
Will remove this, I had added this previously since it was checking for python 2.4 version

bhavikbhavsar replied:
Also will try to add create bug.

jtyr commented on lib/ansible/modules/web_infrastructure/bugzilla_bug.py:
```diff
@@ -0,0 +1,160 @@
+
+    # Constructing the data/parameter for jsonrpc
+    if user and password:
+        data = 'params=[{"Bugzilla_login":"%s","Bugzilla_password":"%s",' \
```
You can format this like this:

```
data = (
    'params=[{"Bugzilla_login":"%s","Bugzilla_password":"%s",'
    '"ids":[%s]}]' % (user, password, bug_id))
```

bhavikbhavsar replied:
noted it's new learning for me.

jtyr commented on lib/ansible/modules/web_infrastructure/bugzilla_bug.py:
```diff
@@ -0,0 +1,160 @@
+    resp, info = bz_get(module, url, data, 'GET')
+
+    if info.get("status") != 200:
+        module.fail_json(msg="HTTP error " +
```
This should be formatted like this:

```
module.fail_json(
    msg="HTTP error " + str(info["status"]) + " " + info["msg"])
```

or like this:

```
module.fail_json(
    msg="HTTP error %s %s" % (info["status"], info["msg"]))
```

bhavikbhavsar replied:
noted

jtyr commented on lib/ansible/modules/web_infrastructure/bugzilla_bug.py:
```diff
@@ -0,0 +1,160 @@
+        module.fail_json(msg=error)
+
+    bug_detail = resp_json
+    module.exit_json(bug=bug_detail)
```
In the `exit_json`, you should also return `changed` and `state`.

bhavikbhavsar replied:
when I am fetching the bug details I am not making changes, so changed will remain false right ?

jtyr reviewed (changes requested):
If the module cannot create bugs, then I don't think it will be ever merged. What you are trying to do (to get bug info) can be done via `get_url` module.

Akasurde commented:
superseded by https://github.com/ansible/ansible/pull/57067

Akasurde commented:
@jtyr @bhavikbhavsar for the contribution.
